rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,316 @@
+
+```yaml
+auth_service:
+  ...
+  authentication:
+    ...
+    private_key_policy: none | hardware_key | hardware_key_touch
+```
+
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  private_key_policy: none | hardware_key | hardware_key_touch
+```
+
+This can also be configured for individual roles:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: role-name
+spec:
+  role_options:
+    private_key_policy: none | hardware_key | hardware_key_touch
+```
+
+- `none` (default): No enforcement on private key usage.
+- `hardware_key`: A user's private keys must be generated on a hardware key. As a result, the user cannot use their signed certificates unless they have their hardware key connected.
+- `hardware_key_touch`: A user's private keys must be generated on a hardware key, and must require touch to be accessed. As a result, the user must touch their hardware key on login, and on every subsequent request.
+
+#### Per-session MFA configuration
+
+```yaml
+auth_service:
+  ...
+  authentication:
+    ...
+    require_session_mfa: on | hardware_key | hardware_key_touch
+```
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  require_session_mfa: on | hardware_key | hardware_key_touch
+```
+```yaml
+kind: role
+version: v5
+metadata:
+  name: role-name
+spec:
+  role_options:
+    require_session_mfa: on | hardware_key | hardware_key_touch
+```
+
+- `on`: Current per-session MFA functionality - user's are required to pass an MFA challenge with a registered MFA device in order to start new SSH|Kubernetes|DB|Desktop sessions. Non-session requests, and app-session requests are not impacted.
+- `hardware_key`: Essentially combines `require_session_mfa: on` with `private_key_policy: true`. MFA tap is only required for session requests, but all Teleport API requests require certificates backed by a hardware private key.
+- `hardware_key_touch`: This is the same as `private_key_policy: hardware_key_touch`. Per-session MFA is disabled in favor of touch checks directly on the hardware key the user used to login.
+
+Note: From a product perspective, we have decided to add the new `require_session_mfa` options and omit the new `private_key_policy` option to reduce configuration knobs. Since the underlying implementation will be completely separate from per-session MFA, this RFD will continue to refer to `private_key_policy` to make reasoning about the proposed changes easier to understand.
+
+#### Enforcement - Attestation
+
+In order to enforce the user private key usage, we need to take a certificate's public key and tie it back to a trusted hardware device, which can be done with attesatation, as explained [above](#attestation).
+
+First, we need to get attestation data from the user's hardware key during login, which we will do with the new rpc `AttestHardwarePrivateKey`.
+
+```proto
+service AuthService {
```

Comment:
I would much prefer for "PIVService" to be separate from "AuthService". It can plug into the same codebase, same port, and same binaries, but it's a small step into breaking the monster that is Auth into more manageable chunks.

Happy to chat about this idea, as I assume this will raise some questions.

(This is not a blocker on my part.)

Reply:
Sure thing, should I follow the layout set by `joinservice.proto` in that case?

Reply:
I would lean towards a separate proto package, but that's alright.